lib/lz4frame.c

```diff
@@ -910,19 +955,29 @@ size_t LZ4F_compressUpdate(LZ4F_cctx* cctxPtr,
                                  srcPtr, (size_t)(srcEnd - srcPtr),
                                  compress, cctxPtr->lz4CtxPtr, cctxPtr->prefs.compressionLevel,
                                  cctxPtr->cdict,
-                                 cctxPtr->prefs.frameInfo.blockChecksumFlag);
-        srcPtr  = srcEnd;
+                                 cctxPtr->prefs.frameInfo.blockChecksumFlag,
+                                 blockCompression);
+      srcPtr = srcEnd;
     }
 
     /* preserve dictionary within @tmpBuff whenever necessary */
     if ((cctxPtr->prefs.frameInfo.blockMode==LZ4F_blockLinked) && (lastBlockCompressed==fromSrcBuffer)) {
-        if (compressOptionsPtr->stableSrc) {
-            cctxPtr->tmpIn = cctxPtr->tmpBuff;  /* src is stable : dictionary remains in src across invocations */
+      if (compressOptionsPtr->stableSrc) {
+        cctxPtr->tmpIn = cctxPtr->tmpBuff;  /* src is stable : dictionary remains in src across invocations */
+      } else {
+        int realDictSize;
+        if (blockCompression == LZ4B_COMPRESSED) {
+          realDictSize = LZ4F_localSaveDict(cctxPtr);
         } else {
-            int const realDictSize = LZ4F_localSaveDict(cctxPtr);
-            assert(0 <= realDictSize && realDictSize <= 64 KB);
-            cctxPtr->tmpIn = cctxPtr->tmpBuff + realDictSize;
+          /* only keep the space of the dictionary, so dict data is kept for the next compressedUpdate
```

Comment:
This portion of the code confuses me.
What is the objective ?

Reply:
The idea was not to write the dictionary if an uncompressed block is written. That's why I've added the get dict size methods. They are used to keep the space of the dictionary free without putting any new data in.
The alternative would be to remove this and always update the dictionary even if we are writing an uncompressed block. 
It would make the dictionary a bit worse but probably simplify this.

Reply:
I'm concerned that this might not be conformant to the frame specification (though I'm unsure if I do understand the details).

Let's quickly state that independent blocks are unaffected, this part is clear.

For linked blocks though, it's specified  that the each block uses previous block(s) as a dictionary.
>  If this flag is set to “0”, each block depends on previous ones (up to LZ4 window size, which is 64 KB). In such case, it’s necessary to decode all blocks in sequence.

Note that each block depends on previous ones, not on previous _compressed_ blocks. This means that, if a block is uncompressed, it's still part of the dictionary for the following block.

I'm not sure how this plays out here.

Reply:
That's why we keep the dictionary but without modifications. The uncompressed block still contains the dictionary but it's not updated with new data. 

lib/lz4frame.c#975
```
    realDictSize = LZ4F_localDictSize(cctxPtr);
}
assert(0 <= realDictSize && realDictSize <= 64 KB);
cctxPtr->tmpIn = cctxPtr->tmpBuff + realDictSize;
````
as real dict size is now set to the last size of the dictionary `cctxPtr->tmpIn` starts behind the dictionary data and the memory of the dict is not modified. When the block is written is still contains the data 

I probably should update the fuzzing test to make sure it's working dependent and independent blocks.

Reply:
Fuzzing test is updated to include both blocked modes.